core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -3734,6 +3740,13 @@ private Exists(RelNode r, boolean indicator, boolean outerJoin) {
       this.outerJoin = outerJoin;
     }
   }
+
+  /** Check if it is the join whose condition is based on column equality. */
+  public static boolean isEquiJoin(Join join) {
+    return join.isNonCorrelateSemiJoin()
+        || join instanceof EnumerableHashJoin
+        || join instanceof EnumerableMergeJoin;
+  }
```

Comment:
This looks weird.

Reply:
Agree, couldn't it simply be: `return join.analyzeCondition().isEqui();`

Reply:
There are 2 rules that use this method to make some decision when planning, this method returns what the original EquiJoin is, cause now the join.analyzeCondition().isEqui() has wider scope that this method returns. I thought maybe we should move the logic to the rules or tweak the rules logic.

Reply:
@hsyuan @rubenada I have rename the method name to forceEquiJoin, cause now some join Enumerable nodes force the join condition must be equi-join, like semi/hash/merge join. So some rule needs this info to prevent some unexpected behavior.

Reply:
@hsyuan @rubenada I have rename the method name to forceEquiJoin, cause now some join Enumerable nodes force the join condition must be equi-join, like semi/hash/merge join. So some rule needs this info to prevent some unexpected behavior.

Reply:
In that case, maybe it would be simpler to not deprecate `EquiJoin` and leave its current sub-classes as extends `EquiJoin` and continue using `instanceof EquiJoin` for this check?

Reply:
@rubenada same as SemiJoin, i just deprecate it, cause the JoinInfo already can describe if the join is a equi-join.

Reply:
> this method returns what the original EquiJoin is, cause now the join.analyzeCondition().isEqui() has wider scope that this method returns. 

What do you mean by wider scope? The semji join is not guaranteed to be an equi join anymore.

Reply:
We should never call enumerable classes here, because enumerable adapter may not be used by other system, most likely.

Reply:
I agree, will keep `EquiJoin` class, and make decision based on inheritance of it.

Reply:
In fact, we don't need EquiJoin at all. The definition of EquiJoin in Calcite is misleading.  
`t1.a+t1.b = t2.c+t2.d` would not be a EquiJoin. What we care about is does it have equal condition that we can do hash or merge join on it. But given the size of change, I don't mind to remove/deprecate EquiJoin in later patches.

Reply:
There is a discussion about what is an `EquiJoin` in the @dev mail list, see [1].

[1] https://lists.apache.org/list.html?dev@calcite.apache.org:lte=1M:What%20is%20the%20exactly%20definition%20as%20an%20equi%20join

Reply:
I would rather trust Oracle's documentation about `EquiJoin`: https://docs.oracle.com/cd/E11882_01/server.112/e41084/queries006.htm#SQLRF52350

However, it is meaningless and pedantic to argue what is an EquiJoin. Think about this question. What makes the so-called `EquiJoin` special? Is it because we can create hash join or merge join for it? If yes, the `EquiJoin` should be removed. As long as there is an equality join condition, both sides of the equality operator contain columns from single table, and there is no subquery in the condition, and both sides are hashable or sortable, even it is not a so-called `EquiJoin`, we can still create hash join or merge join for it. The current `EquiJoin` in Calcite has so many limitations. Even though we can make additional project to make each side become a single column to satisfy `EquiJoin`, it will add another operator, adding additional execution cost.

Reply:
I have add some comments for `EquiJoin` class, we need a class like this to tell Clacite that the join definitely can not have non-equi join conditions, either a tool method or a class, we choose the later, it is not about what is a `EquiJoin` that much, we just need a tag class to let Clacite enumerables and other system to mark that instance.

Reply:
Sorry, I don't get it. What's your plan on `EquiJoin`? Keep it or deprecate it in another patch?

Reply:
I would recommend inline `forceEquiJoin` and remove this function, because it seems like a temporary walk around  for this refactoring. Exposing this as a function makes us hard to remove it in future release.

Reply:
`forceEquiJoin` is just a tool method, i think it is okey cause the logic need to be complete and consistent. For current Calcite, SemiJoin must only have equi-join conditions and we need a method to handle this while `SemiJoin` class is deprecated.

For `EquiJoin` class i would rather to keep it, cause other systems also need it to tell Calcite that their joins can only have equi-join condition predicates.